scipy/sparse/_coo.py

```diff
@@ -524,6 +526,271 @@ def _with_data(self, data, copy=True):
+        accum=np.ones(len(self.data), dtype=np.bool_)
+        slice_coords = []
+        arr_coords = []
+        arr_indices = []
+        for i, (idx, co) in enumerate(zip(index, self.coords)):
+            if isinstance(idx, int):
+                accum &= (co == idx)
+            elif isinstance(idx, slice):
+                if idx == slice(None):
+                    slice_coords.append(co)
+                else:
+                    start, stop, step = idx.indices(self.shape[i])
+                    if step != 1:
+                        if step < 0:
+                            in_range = (co <= start) & (co > stop)
+                        else:
+                            in_range = (co >= start) & (co < stop)
+                        new_ix, m = np.divmod(co - start, step)
+                        accum &= (m == 0) & in_range
+                    else:
+                        in_range = (co >= start) & (co < stop)
+                        new_ix = co - start
+                        accum &= in_range
+                    slice_coords.append(new_ix)
+            else:  # array
+                arr_coords.append(co)
+                arr_indices.append(idx)
+        # shortcut for scalar output
+        if new_shape == ():
+            return self.data[accum].sum().astype(self.dtype, copy=False)
+
+        new_coords = [co[accum] for co in slice_coords]
+        new_data = self.data[accum]
+
+        # handle array indices
+        if arr_indices:
+            arr_shape = arr_indices[0].shape  # already broadcast in validate_indices
+            # keyarr is arr_size by #array indices
+            keyarr = np.array(arr_indices).reshape(len(arr_indices), -1).T
+
+            # arr_coords.T is old_NNZ by #array indices
+            arr_coords = np.array([co[accum] for co in arr_coords])
+            # found is arr_size x old_NNZ. True means coords match all arr_indices
+            found = (keyarr[:, None, :] == arr_coords.T).all(axis=2)
+            arr_ix, arr_co = found.nonzero()
+            new_data = new_data[arr_co]
+            new_coords = [co[arr_co] for co in new_coords]
+            new_arr_coords = list(np.unravel_index(arr_ix, shape=arr_shape))
+
+            # check for contiguous positions of array and int indices
+            if len(arr_int_pos) == arr_int_pos[-1] - arr_int_pos[0] + 1:
+                # Contiguous. Put all array index shape at pos of array indices
+                pos = arr_int_pos[0]
+                new_coords = new_coords[:pos] + new_arr_coords + new_coords[pos:]
+            else:
+                # Not contiguous. Put all array coords at front
+                new_coords = new_arr_coords + new_coords
+
+        if none_pos:
+            if new_coords:
+                coord_like = new_coords[0]
+            else:
+                coord_like = np.zeros(len(new_data), dtype=self.coords[0].dtype)
+            for i in none_pos:
+                new_coords.insert(i, np.zeros_like(coord_like))
```

Comment:
This makes one extra array. We could move the zeros_like to line 593, and then copy on insert `if i != 0`